peer/peer.go

```diff
@@ -0,0 +1,2114 @@
+	getAddrMsgListeners     map[string]func(*Peer, *wire.MsgGetAddr)
+	addrMsgListeners        map[string]func(*Peer, *wire.MsgAddr)
+	pingMsgListeners        map[string]func(*Peer, *wire.MsgPing)
+	pongMsgListeners        map[string]func(*Peer, *wire.MsgPong)
+	alertMsgListeners       map[string]func(*Peer, *wire.MsgAlert)
+	memPoolMsgListeners     map[string]func(*Peer, *wire.MsgMemPool)
+	txMsgListeners          map[string]func(*Peer, *wire.MsgTx)
+	blockMsgListeners       map[string]func(*Peer, *wire.MsgBlock, []byte)
+	invMsgListeners         map[string]func(*Peer, *wire.MsgInv)
+	headersMsgListeners     map[string]func(*Peer, *wire.MsgHeaders)
+	notFoundMsgListeners    map[string]func(*Peer, *wire.MsgNotFound)
+	getDataMsgListeners     map[string]func(*Peer, *wire.MsgGetData)
+	getBlocksMsgListeners   map[string]func(*Peer, *wire.MsgGetBlocks)
+	getHeadersMsgListeners  map[string]func(*Peer, *wire.MsgGetHeaders)
+	filterAddMsgListeners   map[string]func(*Peer, *wire.MsgFilterAdd)
+	filterClearMsgListeners map[string]func(*Peer, *wire.MsgFilterClear)
+	filterLoadMsgListeners  map[string]func(*Peer, *wire.MsgFilterLoad)
+	versionMsgListeners     map[string]func(*Peer, *wire.MsgVersion)
+	rejectMsgListeners      map[string]func(*Peer, *wire.MsgReject)
+}
+
+// String returns the peer's address and directionality as a human-readable
+// string.
+func (p *Peer) String() string {
+	return fmt.Sprintf("%s (%s)", p.addr, directionString(p.inbound))
+}
+
+// isKnownInventory returns whether or not the peer is known to have the passed
+// inventory.  It is safe for concurrent access.
+func (p *Peer) isKnownInventory(invVect *wire.InvVect) bool {
+	p.knownInvMutex.Lock()
+	defer p.knownInvMutex.Unlock()
+
+	if p.knownInventory.Exists(invVect) {
+		return true
+	}
+	return false
+}
+
+// UpdateLastBlockHeight updates the last known block for the peer. It is safe
+// for concurrent access.
+func (p *Peer) UpdateLastBlockHeight(newHeight int32) {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	log.Tracef("Updating last block height of peer %v from %v to %v",
+		p.addr, p.lastBlock, newHeight)
+	p.lastBlock = int32(newHeight)
+}
+
+// UpdateLastAnnouncedBlock updates meta-data about the last block sha this
+// peer is known to have announced. It is safe for concurrent access.
+func (p *Peer) UpdateLastAnnouncedBlock(blkSha *wire.ShaHash) {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	log.Tracef("Updating last blk for peer %v, %v", p.addr, blkSha)
+	p.lastAnnouncedBlock = blkSha
+}
+
+// AddKnownInventory adds the passed inventory to the cache of known inventory
+// for the peer.  It is safe for concurrent access.
+func (p *Peer) AddKnownInventory(invVect *wire.InvVect) {
+	p.knownInvMutex.Lock()
+	defer p.knownInvMutex.Unlock()
+
+	p.knownInventory.Add(invVect)
+}
+
+// AddKnownAddress adds the passed address to the cache of known addresses
+// for the peer.  It is safe for concurrent access.
+func (p *Peer) AddKnownAddress(addr string) {
```

Comment:
I think the addition of the known addresses should be handled by the peer itself before invoking the callback since the filtering it enables is something that should happen for all peers, regardless of the use case.

It would obviously need to be changed slightly from what is currently in the server handler.